R/bibtex.R

```diff
@@ -196,10 +207,93 @@ findBibFile <- function(package) {
 #'
 #' @param file file name
 #' @param encoding encoding
-#' @param srcfile output of \code{\link{srcfile}}
+#' @param srcfile Deprecated
 #' @export
-do_read_bib <- function(file, encoding = "unknown", srcfile){
-  .External( "do_read_bib", file=file, encoding=encoding, srcfile=srcfile, PACKAGE = "bibtex" )
+do_read_bib <- function(file, encoding = "unknown", srcfile) {
+  # Assess the extension of the file
+  if (!file.exists(file)) stop("Error: unable to open file to read")
+
+  # Read all as UTF-8
+  lines <- readLines(file, encoding = encoding, warn = FALSE)
+
+  # Aux
+  trimlines <- trimws(lines)
+
+  # Map bib file
+  # Init and end of citation
+  init <- grep("^@", trimlines)
+
+  # Identify type of entry
+  entry <- unlist(lapply(trimlines[init], function(x) {
+    unlist(strsplit(x, "\\{"))[1]
+  }))
+
+
+  # Map fo document
+  map_bib <- data.frame(
+    entry = trimws(entry),
+    entry_lower = trimws(tolower(entry)),
+    init = init,
+    # The end of each entry is at most the
+    # previous line of the next entry
+    end = c(init[-1] - 1, length(lines))
+  )
+
+  # Extract map string
+  map_string <- map_bib[map_bib$entry_lower == "@string", ]
+
+  if (nrow(map_string) > 0) {
+    # This is needed for multiline strings
+
+    stringlines <- lapply(seq_len(nrow(map_string)), function(x) {
+      init <- map_string$init[x]
+      end <- map_string$end[x]
+      string_line <- lines[seq(init, end)]
+      # Guess lines outside of the entry
+
+      guess_eostring <- max(grep("\\}$", string_line))
+
+      string_line <- string_line[seq(1, guess_eostring)]
+
+      # Collapse to single line
+      string_line <- paste0(string_line, collapse = "\n")
+      string_line
+    })
+
+    stringlines <- unlist(stringlines)
+
+    map_string_end <- tryCatch(parse_strings(map_string, stringlines),
+      warning = function(e) {
+        NULL
+      },
+      error = function(e) {
+        stop("Error when parsing strings of ",
```

Comment:
Is there a way we could indicate which line was problematic? Anything to make the error message more informative would be ideal here.

Reply:
I think we are already doing that, see this reprex. Do you wish to improve it in any way?

``` r
# Message on parsing entries
f <- file.path(system.file("bib", "badFormat.bib", package = "bibtex"))
bib <- read.bib(f)
#> ignoring entry 'brown:1963' (line 6) because :
#>  A bibentry of bibtype 'Book' has to specify the field: title
#> ignoring entry 'chambers+cleveland+kleiner+tukey:1983' (line 21) because :
#>  Invalid author/editor format.

# On unbalanced braces we throw an error indicating the problematic line
f2 <- system.file("bib", "unbalanced_braces.bib", package = "bibtex")
bib2 <- read.bib(f2)
#> Error: Unbalanced braces on entry (line 9). Invalid .bib file
#> Error: Invalid bib file
file <- f2
```

See https://github.com/ropensci/bibtex/blob/05a401cda1d4f1757a82979d181c4486a54755b1/inst/bib/badFormat.bib#L6-L10 and https://github.com/ropensci/bibtex/blob/05a401cda1d4f1757a82979d181c4486a54755b1/inst/bib/unbalanced_braces.bib#L9-L14